### ISSUE TITLE
[Request algo] Stellar XLM Stargazer wallet

### ISSUE THREAD

**Issue by ghost** [hashcat/hashcat#2524]
Is it possible to add bruteforce on hashcat for Stellar XLM Stargazer wallet ? 
I found the solution - script for brute, but only on .JS 

Wallet / https://getstargazer.com/
Script / https://gist.github.com/johansten/35f4767e2f1d4d108a4bb21ff8289ec7/

**Comment by philsmd** [comment]
Please remember that you always should answer all questions / points mentioned here: https://hashcat.net/faq#i_want_to_request_some_new_algorithms_or_features_how_can_i_accomplish_this

It seems that the code works like this:

<details>
  <summary>xlm_stargazer.pl</summary>

```perl
#!/usr/bin/env perl

# Author: philsmd
# Date: August 2020
# License: public domain, credits go to philsmd and hashcat

use strict;
use warnings;

use MIME::Base64 qw (decode_base64);
use Crypt::AuthEnc::GCM;
use Crypt::PBKDF2;

#
# Constants
#

my $ITERATIONS = 4096;

#
# Example
#

# Example 1:

# ["1","ZCtl/+vWiLL358Jz+xnP5A==",["GgmFU37DSX4evSMU","CoMGXWHqDmLwxRAgORqjK/MyFEMAkMbqvDEDMjn4veVwpHab9m6Egcwp70qEJsRhjkHjCMWj9zX40tu9UK5QACuB8gD1r9Cu"]]

my $pass = "lacoin";
my $salt = "ZCtl/+vWiLL358Jz+xnP5A==";
my $IV   = "GgmFU37DSX4evSMU";
my $blob = "CoMGXWHqDmLwxRAgORqjK/MyFEMAkMbqvDEDMjn4veVwpHab9m6Egcwp70qEJsRhjkHjCMWj9zX40tu9UK5QACuB8gD1r9Cu";


#
# Start
#

my $salt_bin = decode_base64 ($salt);
my $IV_bin   = decode_base64 ($IV);
my $blob_bin = decode_base64 ($blob);

my $data = substr ($blob_bin,  0, 56);
my $tag  = substr ($blob_bin, 56, 16); # last 16 bytes make up the random tag

my $pbkdf2 = Crypt::PBKDF2->new
(
  hasher     => Crypt::PBKDF2->hasher_from_algorithm ('HMACSHA2', 256),
  iterations => $ITERATIONS
);

# most demanding part of the algorithm:

my $key = $pbkdf2->PBKDF2 ($salt_bin, $pass);

# AES GCM

my $aes = Crypt::AuthEnc::GCM->new ("AES", $key, $IV_bin);

my $plain_text = $aes->decrypt_add  ($data);
my $result_tag = $aes->decrypt_done ($tag);

if ($result_tag == 1)
{
  print "Password found '$pass'\n";

  exit (0);
}

exit (1);
```

</details>

As the POC (proof of concept) perl code above highlights, the algorithm uses `PBKDF2-HMAC-SHA256` with exactly **4096** iterations hard-coded/fixed.

After that, it uses the derived key for `AES-256 GCM` decryption to verify that the decryption works.
For the GCM mode of operation see: https://en.wikipedia.org/wiki/Galois/Counter_Mode (it's an authenticated algorithm that uses a `tag` to check if the decryption worked correctly).

We currently have no support for `GCM` in hashcat. This is for now a quite uncommon mode of operation for AES, it's very rarely used in algorithms we have/use for password / hash cracking . It would be quite hard/difficult to get `GCM` working because we don't yet have code for this.

Do all the XLM wallet use this ?
Any other more information.

I'm not sure if many users could profit from this new hash mode in hashcat... We should really decide if we really should add all these strange cryptocurrency wallets and algorithms. For now I'm quite reluctant about this... there are so many cryptocurrencies and strange algorithms, sometimes even several wallet implementation with different algorithms for the same cryptocurrency etc etc etc

I know we already support bitcoin/litecoin/ethereum/electrum etc etc etc , but should we really support all of these ? Is this even a common one and a common wallet ? Do all the XLM wallet use this algorithm ?

We should probably at least wait for more user request and thumbs up / reactions here. I'm not too convinced about this yet.

Thank you anyway and maybe next time mention more information about the wallets types and algorithm details / limits etc

**Comment by matrix** [comment]
@kipilk0 @philsmd AES-256-GCM almost done :D

stay tuned ...

**Comment by matrix** [comment]
Done ! :D

i will cleanup/push the code in these days. 

Now the benchmark is:

```
OpenCL API (OpenCL 1.2 (May  7 2020 00:10:14)) - Platform #1 [Apple]
====================================================================
* Device #1: Intel(R) Core(TM) i7-4578U CPU @ 3.00GHz, 8128/8192 MB (2048 MB allocatable), 4MCU
* Device #2: Iris, skipped

Benchmark relevant options:
===========================
* --optimized-kernel-enable

Hashmode: 27000 - Stellar XLM Stargazer Wallet, PBKDF2-HMAC-SHA256 + AES-256-GCM (Iterations: 4095)

Speed.#1.........:     2321 H/s (54.51ms) @ Accel:256 Loops:512 Thr:1 Vec:4
```

Poc

```
$ ./hashcat -m 27000 '$stellar$ZCtl/+vWiLL358Jz+xnP5A==$GgmFU37DSX4evSMU$CoMGXWHqDmLwxRAgORqjK/MyFEMAkMbqvDEDMjn4veVwpHab9m6Egcwp70qEJsRhjkHjCMWj9zX40tu9UK5QACuB8gD1r9Cu' -a3 laco?l?l
[...]
Applicable optimizers applied:
* Zero-Byte
* Single-Hash
* Single-Salt
* Brute-Force
* Slow-Hash-SIMD-LOOP

Watchdog: Hardware monitoring interface not found on your system.
Watchdog: Temperature abort trigger disabled.

Host memory required for this attack: 1 MB

$stellar$ZCtl/+vWiLL358Jz+xnP5A==$GgmFU37DSX4evSMU$CoMGXWHqDmLwxRAgORqjK/MyFEMAkMbqvDEDMjn4veVwpHab9m6Egcwp70qEJsRhjkHjCMWj9zX40tu9UK5QACuB8gD1r9Cu:lacoin
                                                 
Session..........: hashcat
Status...........: Cracked
Hash.Name........: Stellar XLM Stargazer Wallet, PBKDF2-HMAC-SHA256 + AES-256-GCM
Hash.Target......: $stellar$ZCtl/+vWiLL358Jz+xnP5A==$GgmFU37DSX4evSMU$...D1r9Cu
Time.Started.....: Thu Dec 24 20:14:57 2020 (0 secs)
Time.Estimated...: Thu Dec 24 20:14:57 2020 (0 secs)
Guess.Mask.......: laco?l?l [6]
Guess.Queue......: 1/1 (100.00%)
Speed.#1.........:     2336 H/s (6.63ms) @ Accel:16 Loops:1024 Thr:1 Vec:4
Recovered........: 1/1 (100.00%) Digests
Progress.........: 64/676 (9.47%)
Rejected.........: 0/64 (0.00%)
Restore.Point....: 0/676 (0.00%)
Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:3072-4095
Candidates.#1....: lacona -> lacode

```

**Comment by jsteube** [comment]
The hash-mode was merged into hashcat v6.2.0.
